File: C Sharp/Database/SalesByCategory.cs
using System;
using System.Drawing;
using Aspose.Cells;
using Aspose.Cells.Drawing;
using Aspose.Cells.Charts;

namespace Aspose.Cells.Demos
{
    /// <summary>
    /// Summary description for SalesByCategory.
    /// </summary>
    public class SalesByCategory : DbBase
    {
        public SalesByCategory(string path)
            : base(path)
        {
            //
            // TODO: Add constructor logic here
            //
        }

        public Workbook CreateSalesByCategory()
        {
            try
            {
                DBInit();
            }
            catch
            {
            }
            finally
            {
                if (this.oleDbConnection1 != null)
                    this.oleDbConnection1.Close();
            }
            
            //Open the template file
	    string designerFile = MapPath("~/Designer/Northwind.xls");						
            Workbook workbook = new Workbook(designerFile);

            try
            {
                //Specify SQL and execute the query to fill the datatable
                this.oleDbDataAdapter1.SelectCommand.CommandText = @"SELECT DISTINCTROW Categories.CategoryID, 
					Categories.CategoryName, Products.ProductName, SUM([Order Details Extended].ExtendedPrice) AS ProductSales
				FROM  Categories 
				INNER JOIN
					(Products INNER JOIN (Orders INNER JOIN [Order Details Extended] ON
					Orders.OrderID = [Order Details Extended].OrderID) ON Products.ProductID = [Order Details Extended].ProductID) ON Categories.CategoryID = Products.CategoryID
				WHERE
					(((Orders.OrderDate) BETWEEN #1/1/1995# AND 
						#12/31/1995#)) GROUP BY Categories.CategoryID ,  Categories.CategoryName ,  Products.ProductName ORDER BY Categories.CategoryName";
                this.oleDbDataAdapter1.Fill(this.dataTable1);
            }
            catch
            {
            }
            finally
            {
                if (this.oleDbConnection1 != null)
                    this.oleDbConnection1.Close();
            }

            //Get the worksheet
            Worksheet sheet = workbook.Worksheets["Sheet8"];
            //Name the worksheet
            sheet.Name = "Sales By Category";
            //Get the cells
            Cells cells = sheet.Cells;
            //Get the vertical page breaks
            VerticalPageBreakCollection vPageBreaks = sheet.VerticalPageBreaks;
            int currentRow = 2;
            byte currentColumn = 0;

            string lastCategory = "";
            string thisCategory, nextCategory;

            SetSalesByCategoryStyles(workbook);
            //Fill cells with source data and apply styles
            for (int i = 0; i < this.dataTable1.Rows.Count; i++)
            {
                thisCategory = (string)this.dataTable1.Rows[i]["CategoryName"];
                if (thisCategory != lastCategory)
                {
                    currentRow = 2;
                    if (i != 0)
                        currentColumn += 15;
                    CreateSalesByCategoryHeader(workbook, cells, currentRow, currentColumn, thisCategory);
                    lastCategory = thisCategory;
                    currentRow += 2;
                }
                cells[currentRow, currentColumn].PutValue((string)this.dataTable1.Rows[i]["ProductName"]);
                cells[currentRow, (byte)(currentColumn + 1)].PutValue((double)(decimal)this.dataTable1.Rows[i]["ProductSales"]);

                cells[currentRow, (byte)(currentColumn + 1)].SetStyle(workbook.Styles["Sales"]);

                cells.SetColumnWidth(currentColumn, 27);
                cells.SetColumnWidth((byte)(currentColumn + 1), 15);

                if (i != this.dataTable1.Rows.Count - 1)
                {
                    nextCategory = (string)this.dataTable1.Rows[i + 1]["CategoryName"];
                    if (thisCategory != nextCategory)
                    {
                        vPageBreaks.Add(0, currentColumn + 1);
                        CreateChart(workbook, sheet, currentRow, currentColumn);
                    }
                }
                else
                {
                    CreateChart(workbook, sheet, currentRow, currentColumn);
                }
                currentRow++;
            }
            //Remove the unnecessary worksheets
            for (int i = 0; i < workbook.Worksheets.Count; i++)
            {
                sheet = workbook.Worksheets[i];
                if (sheet.Name != "Sales By Category")
                {
                    workbook.Worksheets.RemoveAt(i);
                    i--;
                }
            }
            //Get the workbook (generated)
            return workbook;
        }

        private void CreateChart(Workbook workbook, Worksheet sheet, int currentRow, int currentColumn)
        {
            //Add a bar chart
            int chartIndex = sheet.Charts.Add(ChartType.Bar, 4, currentColumn + 3,
                26, currentColumn + 14);
            //Get the chart
            Chart chart = sheet.Charts[chartIndex];
            //Make the legends invisible
            chart.ShowLegend = false;
            string startCell = CellsHelper.CellIndexToName(4, currentColumn + 1);
            string endCell = CellsHelper.CellIndexToName(currentRow, currentColumn + 1);
            //Set the nseries
            chart.NSeries.Add(startCell + ":" + endCell, true);
            //Set the fill format for the plot area
            FillFormat fillFormat = chart.PlotArea.Area.FillFormat;
            fillFormat.SetPresetColorGradient(GradientPresetType.Daybreak, GradientStyleType.Vertical, 1);
            //Set the category data
            startCell = CellsHelper.CellIndexToName(4, currentColumn);
            endCell = CellsHelper.CellIndexToName(currentRow, currentColumn);
            chart.NSeries.CategoryData = startCell + ":" + endCell;
        }
        private void CreateSalesByCategoryHeader(Workbook workbook, Cells cells, int currentRow, byte currentColumn, string categoryName)
        {
            //Input data and apply style
            cells[currentRow, currentColumn].PutValue(categoryName);
            cells[currentRow, currentColumn].SetStyle( workbook.Styles["Header"]);
        }

        private void SetSalesByCategoryStyles(Workbook workbook)
        {
            //Create a style with specific formatting attributes
            int styleIndex = workbook.Styles.Add();
            Style style = workbook.Styles[styleIndex];
            style.Number = 7;
            style.Name = "Sales";

            //Create a style withe specific set of attributes
            styleIndex = workbook.Styles.Add();
            style = workbook.Styles[styleIndex];
            style.Font.Size = 14;
            style.Font.IsBold = true;
            style.Font.IsItalic = true;
            style.Font.Color = Color.Yellow;
            style.ForegroundColor = Color.Blue;
            style.Pattern = BackgroundType.Solid;
            style.Name = "Header";
        }

    }
}

File: Visual Basic/Database/SalesByCategory.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports Aspose.Cells
Imports Aspose.Cells.Drawing
Imports Aspose.Cells.Charts

Namespace Aspose.Cells.Demos
	''' <summary>
	''' Summary description for SalesByCategory.
	''' </summary>
	Public Class SalesByCategory
		Inherits DbBase
		Public Sub New(ByVal path As String)
			MyBase.New(path)
			'
			' TODO: Add constructor logic here
			'
		End Sub

		Public Function CreateSalesByCategory() As Workbook
			Try
				DBInit()
			Catch
			Finally
				If Me.oleDbConnection1 IsNot Nothing Then
					Me.oleDbConnection1.Close()
				End If
			End Try

			'Open the template file
		Dim designerFile As String = MapPath("~/Designer/Northwind.xls")
			Dim workbook As New Workbook(designerFile)

			Try
				'Specify SQL and execute the query to fill the datatable
				Me.oleDbDataAdapter1.SelectCommand.CommandText = "SELECT DISTINCTROW Categories.CategoryID, " & ControlChars.CrLf & "					Categories.CategoryName, Products.ProductName, SUM([Order Details Extended].ExtendedPrice) AS ProductSales" & ControlChars.CrLf & "				FROM  Categories " & ControlChars.CrLf & "				INNER JOIN" & ControlChars.CrLf & "					(Products INNER JOIN (Orders INNER JOIN [Order Details Extended] ON" & ControlChars.CrLf & "					Orders.OrderID = [Order Details Extended].OrderID) ON Products.ProductID = [Order Details Extended].ProductID) ON Categories.CategoryID = Products.CategoryID" & ControlChars.CrLf & "				WHERE" & ControlChars.CrLf & "					(((Orders.OrderDate) BETWEEN #1/1/1995# AND " & ControlChars.CrLf & "						#12/31/1995#)) GROUP BY Categories.CategoryID ,  Categories.CategoryName ,  Products.ProductName ORDER BY Categories.CategoryName"
				Me.oleDbDataAdapter1.Fill(Me.dataTable1)
			Catch
			Finally
				If Me.oleDbConnection1 IsNot Nothing Then
					Me.oleDbConnection1.Close()
				End If
			End Try

			'Get the worksheet
			Dim sheet As Worksheet = workbook.Worksheets("Sheet8")
			'Name the worksheet
			sheet.Name = "Sales By Category"
			'Get the cells
			Dim cells As Cells = sheet.Cells
			'Get the vertical page breaks
			Dim vPageBreaks As VerticalPageBreakCollection = sheet.VerticalPageBreaks
			Dim currentRow As Integer = 2
			Dim currentColumn As Byte = 0

			Dim lastCategory As String = ""
			Dim thisCategory, nextCategory As String

			SetSalesByCategoryStyles(workbook)
			'Fill cells with source data and apply styles
			For i As Integer = 0 To Me.dataTable1.Rows.Count - 1
				thisCategory = CStr(Me.dataTable1.Rows(i)("CategoryName"))
				If thisCategory <> lastCategory Then
					currentRow = 2
					If i <> 0 Then
						currentColumn += 15
					End If
					CreateSalesByCategoryHeader(workbook, cells, currentRow, currentColumn, thisCategory)
					lastCategory = thisCategory
					currentRow += 2
				End If
				cells(currentRow, currentColumn).PutValue(CStr(Me.dataTable1.Rows(i)("ProductName")))
				cells(currentRow, CByte(currentColumn + 1)).PutValue(CDbl(CDec(Me.dataTable1.Rows(i)("ProductSales"))))

				cells(currentRow, CByte(currentColumn + 1)).SetStyle(workbook.Styles("Sales"))

				cells.SetColumnWidth(currentColumn, 27)
				cells.SetColumnWidth(CByte(currentColumn + 1), 15)

				If i <> Me.dataTable1.Rows.Count - 1 Then
					nextCategory = CStr(Me.dataTable1.Rows(i + 1)("CategoryName"))
					If thisCategory <> nextCategory Then
						vPageBreaks.Add(0, currentColumn + 1)
						CreateChart(workbook, sheet, currentRow, currentColumn)
					End If
				Else
					CreateChart(workbook, sheet, currentRow, currentColumn)
				End If
				currentRow += 1
			Next i
			'Remove the unnecessary worksheets
            Dim iworkbook As Integer = 0
            Do While iworkbook < workbook.Worksheets.Count
                sheet = workbook.Worksheets(iworkbook)
                If sheet.Name <> "Sales By Category" Then
                    workbook.Worksheets.RemoveAt(iworkbook)
                    iworkbook -= 1
                End If
                iworkbook += 1
            Loop
			'Get the workbook (generated)
			Return workbook
		End Function

		Private Sub CreateChart(ByVal workbook As Workbook, ByVal sheet As Worksheet, ByVal currentRow As Integer, ByVal currentColumn As Integer)
			'Add a bar chart
			Dim chartIndex As Integer = sheet.Charts.Add(ChartType.Bar, 4, currentColumn + 3, 26, currentColumn + 14)
			'Get the chart
			Dim chart As Chart = sheet.Charts(chartIndex)
			'Make the legends invisible
			chart.ShowLegend = False
			Dim startCell As String = CellsHelper.CellIndexToName(4, currentColumn + 1)
			Dim endCell As String = CellsHelper.CellIndexToName(currentRow, currentColumn + 1)
			'Set the nseries
			chart.NSeries.Add(startCell & ":" & endCell, True)
			'Set the fill format for the plot area
			Dim fillFormat As FillFormat = chart.PlotArea.Area.FillFormat
			fillFormat.SetPresetColorGradient(GradientPresetType.Daybreak, GradientStyleType.Vertical, 1)
			'Set the category data
			startCell = CellsHelper.CellIndexToName(4, currentColumn)
			endCell = CellsHelper.CellIndexToName(currentRow, currentColumn)
			chart.NSeries.CategoryData = startCell & ":" & endCell
		End Sub
		Private Sub CreateSalesByCategoryHeader(ByVal workbook As Workbook, ByVal cells As Cells, ByVal currentRow As Integer, ByVal currentColumn As Byte, ByVal categoryName As String)
			'Input data and apply style
			cells(currentRow, currentColumn).PutValue(categoryName)
			cells(currentRow, currentColumn).SetStyle(workbook.Styles("Header"))
		End Sub

		Private Sub SetSalesByCategoryStyles(ByVal workbook As Workbook)
			'Create a style with specific formatting attributes
			Dim styleIndex As Integer = workbook.Styles.Add()
			Dim style As Style = workbook.Styles(styleIndex)
			style.Number = 7
			style.Name = "Sales"

			'Create a style withe specific set of attributes
			styleIndex = workbook.Styles.Add()
			style = workbook.Styles(styleIndex)
			style.Font.Size = 14
			style.Font.IsBold = True
			style.Font.IsItalic = True
			style.Font.Color = Color.Yellow
			style.ForegroundColor = Color.Blue
			style.Pattern = BackgroundType.Solid
			style.Name = "Header"
		End Sub

	End Class
End Namespace


